firebase/test/core/data/gateways/firestore-gateways.spec.ts

```diff
@@ -0,0 +1,112 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import * as assert from "assert";
+import * as sinon from "sinon";
+import { FirestoreGateway, QueryFilter } from "@data/gateways/firestore-gateway";
+import * as firebase from "firebase-admin";
+import createSinonStub from "@test/sinon-stub";
+import FirebaseFirestoreError from "@faults/errors/firebase-firestore-error";
+
+// TODO(matuella): Improve readability:
+// 1. Can't export type alias, meaning that it's really verbose to write related types.
+// 2. Improve type-narrowing. Using "as any" as a workaround in a couple of cases.
+
+describe("FirestoreGateway", () => {
+  type FakeCollection = "collection";
+
+  let firestoreMock: sinon.SinonStubbedInstance<firebase.firestore.Firestore>;
+  let firestoreGateway: FirestoreGateway<FakeCollection>;
+
+  beforeEach(() => {
+    firestoreMock = createSinonStub(firebase.firestore.Firestore);
+    firestoreGateway = new FirestoreGateway<FakeCollection>(firestoreMock as firebase.firestore.Firestore);
+  });
+
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  function createTransactionMock(): sinon.SinonStubbedInstance<firebase.firestore.Transaction> {
+    const transactionMock = createSinonStub(firebase.firestore.Transaction);
+    firestoreMock.runTransaction.callsFake((fn) => {
+      return fn(transactionMock);
+    });
+
+    return transactionMock;
+  }
+
+  const fakeObject = { test: "object" };
+
+  describe("read", () => {
+    let querySnapshotMock: sinon.SinonStubbedInstance<firebase.firestore.QuerySnapshot>;
+    let docDataMock: sinon.SinonStubbedInstance<firebase.firestore.DocumentSnapshot>;
+
+    function createQuerySnapshotMock() {
+      const querySnapshotMock = createSinonStub(firebase.firestore.QuerySnapshot);
+      sinon.replaceGetter(querySnapshotMock, "docs", () => [docDataMock] as firebase.firestore.QueryDocumentSnapshot[]);
+      return querySnapshotMock;
```

Comment:
Why can't this one be equal to the below? Like:

```suggestion
      const querySnapshotMock = createSinonStub(firebase.firestore.QuerySnapshot);
      querySnapshotMock.docs.returns([docDataMock])
      
      return querySnapshotMock;
```

Reply:
Because it's a property and not a method